File: Java/dependency-injection/_05_service-locator/src/main/java/ServiceLocatorAntipattern/Core/AcmeProcessingWorkflow.java
package ServiceLocatorAntipattern.Core;

import ServiceLocatorAntipattern.ApplicationRoot;
import ServiceLocatorAntipattern.Interfaces.Message;
import ServiceLocatorAntipattern.Outbound.Outbound;
import ServiceLocatorAntipattern.Services.Authorization;
import ServiceLocatorAntipattern.Services.Repository;

public class AcmeProcessingWorkflow implements ProcessingWorkflow {
  private final Repository repository;
  private final Authorization authorizationRules;
  private Outbound outbound;

  public AcmeProcessingWorkflow() {
    authorizationRules = ApplicationRoot.context.getComponent(Authorization.class);
    repository = ApplicationRoot.context.getComponent(Repository.class);
  }

  public void setOutbound(Outbound outbound) {
    this.outbound = outbound;
  }

  public void applyTo(Message message) {
    message.authorizeUsing(authorizationRules);
    repository.save(message);
    outbound.send(message);
  }
}

File: CSharp/DependencyInjection/DependencyInjection/ServiceLocatorDIAntipattern/_2_ServiceLocator/Core/AcmeProcessingWorkflow.cs
using ServiceLocatorDIAntipattern._2_ServiceLocator.Interfaces;
using ServiceLocatorDIAntipattern._2_ServiceLocator.Outbound;
using ServiceLocatorDIAntipattern._2_ServiceLocator.Services;
using Unity;

namespace ServiceLocatorDIAntipattern._2_ServiceLocator.Core
{
  internal interface IProcessingWorkflow
  {
    void SetOutbound(IOutbound outbound);
    void ApplyTo(AcmeMessage message);
  }

  class AcmeProcessingWorkflow : IProcessingWorkflow
  {
    private readonly IRepository _repository;
    private readonly IAuthorization _authorizationRules;
    private IOutbound _outbound;

    public AcmeProcessingWorkflow()
    {
      _authorizationRules = ApplicationRoot.Context.Resolve<IAuthorization>();
      _repository = ApplicationRoot.Context.Resolve<IRepository>();
    }

    public void SetOutbound(IOutbound outbound)
    {
      _outbound = outbound;
    }

    public void ApplyTo(AcmeMessage message)
    {
      message.AuthorizeUsing(_authorizationRules);
      _repository.Save(message);
      _outbound.Send(message);
    }
  }
}